pkg/core/edge_kb_test.go

```diff
@@ -0,0 +1,232 @@
+				{{typeB, typeC}, {typeC, typeD}, {typeD, typeE}},
+				{{typeB, typeC}, {typeC, typeE}},
+				{{typeA, typeB}, {typeA, typeE}},
+			},
+		},
+		{
+			name:   "paths from d to c",
+			source: &D{},
+			dest:   &C{},
+			want: [][]Edge{
+				{{typeD, typeB}, {typeB, typeC}},
+			},
+		},
+		{
+			name:   "no paths from e",
+			source: &E{},
+			dest:   &A{},
+			want:   [][]Edge{},
+		},
+	}
+	for _, tt := range cases {
+		t.Run(tt.name, func(t *testing.T) {
+			fmt.Println(tt.name)
+			assert := assert.New(t)
+
+			result := FindPaths(TestKnowledgeBase, reflect.TypeOf(tt.source), reflect.TypeOf(tt.dest))
+			assert.ElementsMatch(tt.want, result)
+		})
+	}
+}
+
+func Test_ConfigureFromEdgeData(t *testing.T) {
+	cases := []struct {
+		name   string
+		source Resource
+		dest   Resource
+		data   EdgeData
+		want   []klothograph.Edge[Resource]
+	}{
+		{
+			name:   "node must and must not exist",
+			source: &A{},
+			dest:   &B{},
+			want: []klothograph.Edge[Resource]{
+				{Source: &A{Name: "name"}, Destination: &B{}, Properties: graph.EdgeProperties{Attributes: map[string]string{}, Data: EdgeData{}}},
+			},
+		},
+	}
+	for _, tt := range cases {
+		t.Run(tt.name, func(t *testing.T) {
+			assert := assert.New(t)
+			dag := NewResourceGraph()
+			dag.AddDependencyWithData(tt.source, tt.dest, tt.data)
+			err := ConfigureFromEdgeData(TestKnowledgeBase, dag)
+			assert.NoError(err)
+			assert.ElementsMatch(tt.want, dag.ListDependencies())
+		})
+	}
+}
+
+func Test_ExpandEdges(t *testing.T) {
+	cases := []struct {
+		name   string
+		source Resource
+		dest   Resource
+		data   EdgeData
+		want   []klothograph.Edge[Resource]
+	}{
+		{
+			name:   "node must and must not exist",
+			source: &A{Name: "A"},
+			dest:   &E{Name: "E"},
+			data: EdgeData{
+				Constraint: EdgeConstraint{
+					NodeMustExist:    []Resource{&C{}},
+					NodeMustNotExist: []Resource{&D{}},
+				},
+			},
+			want: []klothograph.Edge[Resource]{
+				{Source: &A{Name: "A"}, Destination: &B{Name: "B"}, Properties: graph.EdgeProperties{Attributes: map[string]string{}}},
+				{Source: &B{Name: "B"}, Destination: &C{}, Properties: graph.EdgeProperties{Attributes: map[string]string{}}},
+				{Source: &C{}, Destination: &E{Name: "E"}, Properties: graph.EdgeProperties{Attributes: map[string]string{}}},
+			},
+		},
+	}
+	for _, tt := range cases {
+		t.Run(tt.name, func(t *testing.T) {
+			assert := assert.New(t)
+			dag := NewResourceGraph()
+			dag.AddDependencyWithData(tt.source, tt.dest, tt.data)
+			err := ExpandEdges(TestKnowledgeBase, dag)
+			assert.NoError(err)
+			assert.ElementsMatch(tt.want, dag.ListDependencies())
+		})
+	}
+}
+
+type (
+	A struct {
+		Name string
+	}
+	B struct {
+		Name string
+	}
+	C struct {
+		Name string
+	}
+	D struct {
+		Name string
+	}
+	E struct {
+		Name string
+	}
+)
+
+func (f *A) Id() ResourceId                      { return ResourceId{Name: "A" + f.Name} }
+func (f *A) Provider() string                    { return "DummyProvider" }
```

Comment:
This isn't on the `Resource` interface anymore (should be in the `Id` function instead)

Reply:
oh nice, i copy pasted from other tests so didnt remove, but removing now